src/worker/activities/activity_heartbeat_manager.rs

```diff
@@ -105,143 +111,221 @@ impl ActivityHeartbeatManager {
+                incoming_hbs,
+            },
+            heartbeat_tx,
+        )
+    }
+
+    /// Record a heartbeat received from lang
+    fn record(&mut self, hb: ValidActivityHeartbeat) -> Option<HeartbeatExecutorAction> {
+        match self.tt_to_state.entry(hb.task_token.clone()) {
+            Entry::Vacant(e) => {
+                let state = ActivityHeartbeatState {
+                    task_token: hb.task_token.clone(),
+                    throttle_interval: hb.throttle_interval,
+                    last_send_requested: Instant::now(),
+                    last_recorded_details: None,
+                    sleep_cancel_tx: None,
+                };
+                e.insert(state);
+                Some(HeartbeatExecutorAction::Report(hb.task_token, hb.details))
+            }
+            Entry::Occupied(mut o) => {
+                let state = o.get_mut();
+                state.last_recorded_details = Some(hb.details);
+                None
+            }
+        }
+    }
+
+    /// Heartbeat report to server completed
+    fn complete_report(&mut self, tt: TaskToken) -> Option<HeartbeatExecutorAction> {
+        if let Some(st) = self.tt_to_state.get_mut(&tt) {
+            let cancellation_token = self.cancellation_token.child_token();
+            st.sleep_cancel_tx = Some(cancellation_token.clone());
+            // Always sleep for simplicity even if the duration is 0
+            Some(HeartbeatExecutorAction::Sleep(
+                tt.clone(),
+                st.get_throttle_sleep_duration(),
+                cancellation_token,
+            ))
+        } else {
+            None
+        }
+    }
+
+    /// Throttling completed, report or stop tracking task token
+    fn complete_throttle(&mut self, tt: TaskToken) -> Option<HeartbeatExecutorAction> {
+        match self.tt_to_state.entry(tt.clone()) {
+            Entry::Occupied(mut e) => {
+                let state = e.get_mut();
+                if state.last_recorded_details.is_none() {
+                    // Nothing to report, forget this task token
+                    e.remove();
+                    None
+                } else {
+                    // Reset the cancellation channel and schedule another report
+                    state.sleep_cancel_tx = None;
+                    state.last_send_requested = Instant::now();
+                    state
+                        .last_recorded_details
+                        .take() // Delete the recorded details before reporting
+                        .map(|details| HeartbeatExecutorAction::Report(tt, details))
+                }
+            }
+            Entry::Vacant(_) => None,
+        }
+    }
+
+    /// Activity should not be tracked anymore, cancel throttle timer if running
+    fn evict(&mut self, tt: TaskToken) -> Option<HeartbeatExecutorAction> {
+        if let Some(state) = self.tt_to_state.remove(&tt) {
+            if let Some(tx) = state.sleep_cancel_tx {
```

Comment:
Could collapse branches by using `.map(|state| state.sleep_cancel_tx)`

Reply:
Tried it but it made the code less readable imho especially because `sleep_cancel_tx` is an `Option`.